executor: support global kill (#17649)
pingcap/tidb: docs/design/2020-06-01-global-kill.md

1. Global connection ids, which are unique amount all TiDB instances.
2. Redirect the `KILL x` to the proper TiDB instance, on which the connection `x` is running.

## Rationale

1. Assign a `server_id` to each TiDB instance on startup by PD, start from 1.
2. Extend connection id to 64 bits, which is composed by `server_id` in high-order 32 bits, and local connection id in low-order 32 bits which is allocated by each TiDB instance itself.
3. On processing `KILL x` command, first extract `server_id` from `x`. If `server_id` aims to a remote TiDB instance, get address from cluster info, and then redirect the command to it by "MySQL API", along with the original user authentication.
4. As some clients would send `KILL x` command with truncated 32 bits connection ids, if `server_id` equals `0`, we should deal with this circumstance as [what we do](https://pingcap.com/docs/stable/sql-statements/sql-statement-kill/) in older versions, i.e, run on current instance if `compatible-kill-query = true`, otherwise, reject it.

## Compatibility

1. Extend connection ids from 32 to 64 bits would break some clients. Main current clients need to be checked for compatibility.

> 2. Some clients would probably send `KILL` command with 32 bits connection ids. We should deal with this circumstance. (Ref. [1](https://github.com/pingcap/tidb/issues/8854#issuecomment-637217000) [2](https://github.com/pingcap/tidb/issues/8854#issuecomment-637237183))

Reviewer: I suggest to leave some space discuss about mysql-client Ctrl+C compatibility. We don't need to give solutions in this RFC and we will resolve it later by introducing another RFC specifically for further introducing a 32 bit global connection id for small clusters.

                                 |             ???             | markup |
                                 |             ???             |(1b,==0)|
                                 +-----------------------------+--------+
```

#### 2. markup
-  `markup == 0` indicates that the `connID` is just 32 bits long effectively, and high 32 bits should be all zeros. Compatible with legacy 32 bits clients.
-  `markup == 1` indicates that the `connID` is 64 bits long. Incompatible with legacy 32 bits clients.
-  `markup == 1` while __high 32 bits are zeros__, indicates that 32 bits truncation happens. See `Compatibility` section.

#### 3. serverId
`serverId` is selected RANDOMLY by each TiDB instance on startup, and the uniqueness is guaranteed by PD(etcd). `serverId` should be larger or equal to 1, to insure that high 32 bits of `connId` should always be non-zero, and make it possible to detect truncation.

On failure (e.g. fail to connect to PD, or all `serverId` are unavailable), we set `serverId` to `0`. And when `serverId == 0`, we deal with `KILL x` as in [early versions](https://pingcap.com/docs/stable/sql-statements/sql-statement-kill/).

> `serverId` is kept by PD with a lease default to 1 hour. If TiDB is disconnected to PD longer than half of the lease, all connections are killed. On connection to PD restored, a new `serverId` is acquired.

Author: I'm sorry for forgetting to update document after codes modified. Comments addressed, PTAL~

On single TiDB instance without PD, a `serverId` of `1` is assigned.

#### 4. local connId
`local connId` is allocated by each TiDB instance on establishing connections incrementally.

Integer overflow is ignored at this stage, as `local connId` should be long enough.

#### 5. global kill

> On processing `KILL x` command, first extract `serverId` from `x`. Then if `serverId` aims to a remote TiDB instance, get the address from cluster info, and redirect the command to it by "Coprocessor API" provided by the remote TiDB, along with the original user authentication.

Reviewer 2: > get the address from cluster info, and redirect the command to it

Is it a broadcasted Coprocessor by now?

+-----------------------+--------------------------------------+--------+
```
##### 32 bits version
(To be discussed in another RFC)
```
                                  31                          1    0
                                 +-----------------------------+--------+
                                 |             ???             | markup |
                                 |             ???             |(1b,==0)|
                                 +-----------------------------+--------+
```

#### 2. markup
-  `markup == 0` indicates that the `connID` is just 32 bits long effectively, and high 32 bits should be all zeros. Compatible with legacy 32 bits clients.
-  `markup == 1` indicates that the `connID` is 64 bits long. Incompatible with legacy 32 bits clients.
-  `markup == 1` while __high 32 bits are zeros__, indicates that 32 bits truncation happens. See `Compatibility` section.

#### 3. serverId
`serverId` is selected RANDOMLY by each TiDB instance on startup, and the uniqueness is guaranteed by PD(etcd). `serverId` should be larger or equal to 1, to insure that high 32 bits of `connId` should always be non-zero, and make it possible to detect truncation.

> On single TiDB instance without PD, a `serverId` of `0` is assigned.

Reviewer 2: What will happen if TiDB runs with mocktikv?

Author: On single TiDB instance without PD, a `serverId` of `1` is assigned. (I have corrected the design doc.)
And as a result, TiDB running with mocktikv behaves just the same as an instances running with PD & TiKV.

> When `serverId == 0`, we deal with `KILL x` as in [early versions](https://pingcap.com/docs/stable/sql-statements/sql-statement-kill/).

Reviewer: server id = 0 means it is a standalone TiDB without PD, then we can safely kill local connections?

Author: Server id = 0 will happen in other occasion, e.g. retry times of acquiring server id from PD exceeds maximum limits.
So shall we assign a special id (e.g. 1) on instance without PD, to enable simple KILL ?

Reviewer: LGTM!

`serverId` is kept by PD with a lease default to 1 hour. If TiDB is disconnected to PD longer to half of the lease, all connections are killed. On connection to PD restored, a new `serverId` is acquired.

#### 4. local connId

> `local connId` is allocated by each TiDB instance on establishing connections.

Reviewer: ```suggestion
`local connId` is allocated by each TiDB instance on establishing connections incrementally.
```

> Integer overflow is ignored at this stage, as `local connId` should be long enouth.

Reviewer 2: ```suggestion
Integer overflow is ignored at this stage, as `local connId` should be long enough.
```

#### 5. global kill

> On processing `KILL x` command, first extract `serverId` from `x`. Then if `serverId` aims to a remote TiDB instance, get address from cluster info, and redirect the command to it by "MySQL API", along with the original user authentication.

Reviewer 2: ```suggestion
On processing `KILL x` command, first extract `serverId` from `x`. Then if `serverId` aims to a remote TiDB instance, get the address from cluster info, and redirect the command to it by "Coprocerssor API" provided by the remote TiDB, along with the original user authentication.
```

## Compatibility

> - Incompatible with legacy 32 bits clients. (According to some quick tests by now, MySQL client v8.0.19 supports `KILL` a connection with 64 bits `connId`, while `CTRL-C` does not, because it truncates the `connId` to 32 bits.)

Reviewer 2: How about telling the fact that if an older version MySQL client triggers this operation?
The detailed behavior should be described with an old client or old SDK.

> - TiDB is able to probe the happening of 32 bits truncation, and deal with a truncated `connId` as [what we do](https://pingcap.com/docs/stable/sql-statements/sql-statement-kill/) in early versions, i.e, run on current instance with `local connId` if `compatible-kill-query = true`, otherwise reject it to prevent killing a wrong connection.

Reviewer 2: Should we keep `compatible-kill-query` config after this PR?

Author: TiDB will degrade to compatible kill when fail to acquire available server id from PD. But this is a rare condition. So I think we can delete `compatible-kill-query` config, but keep `KILL TIDB x` ?
